lib/services/unified_resource.go

```diff
@@ -998,6 +969,160 @@ type resource interface {
+		isResourceServer:     c.isResourceServer,
+		supportsHealthStatus: c.supportsHealthStatus,
+		servers:              maps.Clone(c.servers),
+		sortKey:              c.sortKey,
+	}
+}
+
+type serverWithHealthStatus struct {
+	server resourceServer
+	status types.TargetHealthStatus
+}
+
+type resourceServer interface {
+	resource
+	GetHostID() string
+}
+
+func newResourceCollection(kind string, sortKey resourceSortKey) *resourceCollection {
+	return &resourceCollection{
+		sortKey:              sortKey,
+		isResourceServer:     isResourceServer(kind),
+		supportsHealthStatus: isTargetHealthGetter(kind),
+		servers:              make(map[string]serverWithHealthStatus),
+	}
+}
+
+func isResourceServer(kind string) bool {
+	switch kind {
+	case types.KindAppServer,
+		types.KindDatabaseServer,
+		types.KindKubeServer,
+		types.KindWindowsDesktop:
+		return true
+	default:
+		return false
+	}
+}
+
+func isTargetHealthGetter(kind string) bool {
+	switch kind {
+	case types.KindDatabaseServer:
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *resourceCollection) put(r resource) {
+	c.latest = r
+	if !c.isResourceServer {
+		return
+	}
+	if rs, ok := r.(resourceServer); ok {
+		id := rs.GetHostID()
+		server := serverWithHealthStatus{
+			server: rs,
+		}
+		if r, ok := r.(types.TargetHealthGetter); ok {
+			server.status = types.TargetHealthStatus(r.GetTargetHealth().Status)
+		}
+		c.servers[id] = server
+	}
+}
+
+func (c *resourceCollection) healthStatuses() []types.TargetHealthStatus {
+	if !c.supportsHealthStatus {
+		return nil
+	}
+	var out []types.TargetHealthStatus
+	for _, s := range c.servers {
+		out = append(out, s.status)
+	}
+	return out
+}
+
+func (c *resourceCollection) removeLocked(r types.Resource) {
+	// This looks insane, but we only get a [types.ResourceHeader] in
+	// [types.OpDelete] events.
+	// The types that actually implement [resourceServer] all store the host ID
+	// in the description of the resource header metadata on deletion.
+	// If a new type is added that implements [resourceServer] and the
+	// unified resource watchers starts watching it, then please:
+	// - add it to the isResourceServer helper func
+	// - ensure host ID is stored in the metadata description in its event parser
+	// - add test coverage for it in TestUnifiedResourceWatcher_DeleteEvent
+	if c.isResourceServer {
+		delete(c.servers, r.GetMetadata().Description)
+		return
+	}
+	// Otherwise, the default behavior is the old behavior prior to the
+	// introduction of [resourceCollection]: delete the entire item.
+	clear(c.servers)
+}
+
+func (c *resourceCollection) isSingular() bool {
+	return len(c.servers) <= 1
+}
+
+func (c *resourceCollection) getKind() string {
+	return c.latest.GetKind()
+}
+
+func (c *resourceCollection) match(fn resourceMatchFunc) (resource, bool, error) {
+	if c.isSingular() {
+		match, err := fn(c.latest)
+		return c.latest, match, trace.Wrap(err)
+	}
+
+	for _, r := range c.servers {
```

Comment:
we should think this through, and define clearly what's the behavior when using predicate on health status for an unified resource.

for example, if one agent is healthy but the other is not, and you search for `health.status == "healthy"`, you would see a tile with warning shows up.

Personally, i am still bias towards a calculated single status that aggregates from all heartbeats. Let me know what you think.

Reply:
> I think that makes sense - you search for healthy and there is a healthy server for that db, we just group them all together in one tile. You see a warning because not all servers for that db are healthy, which also seems to make sense

Do we show the healthy server on UI? the side panel only show problematic host IDs. it would be confusing to the user why it showed up.

Reply:
the match function is still against individual servers or should it against the aggregated status? if later we can also simplify the copying logic